New CMORizers for datasets from the NCEP family

hb326 opened the pull request:
## Description
We added three CMORizers for reanalyses from the NCEP family: an update/replacement for the NCEP CMORizer written in ncl (now written in Python), an NCEP-DOE-R2 CMORizer, and a NOAA-CIRES-20CR CMORizer.
This pull request is related to #2864.
Input.rst is updated and extended to reflect the changes.

* * *

- [x] [☝ Create an issue](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#contributing-code-and-documentation) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review


### [New or updated data reformatting script](https://docs.esmvaltool.org/en/latest/develop/dataset.html)

- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/dataset.html#dataset-documentation) is available
- [x] [🛠][1] The dataset has been [added to the CMOR check recipe](https://docs.esmvaltool.org/en/latest/community/dataset.html#testing)
- [x] [🧪][2] Numbers and units of the data look [physically meaningful](https://docs.esmvaltool.org/en/latest/community/dataset.html#scientific-sanity-check)

***

To help with the number of pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

valeriupredoi commented:
guys @hb326 @LisaBock - this is a very large PR (1000 lines of total code change) - could I be a bother and please ask you guys to produce smaller PRs that are more digestible (eg one CMORizer at a time). Having done my bit of complaining, I shall review the technical bits (Python code) and it'd be great if @remi-kazeroni had time to do the data and standards dance :man_dancing:

hb326 commented:
@valeriupredoi, we had thought about splitting the PR up in two (or three) different parts, but decided against it since the main part of this PR is the new cmorizer for NCEP-NCAR-R1, and the cmorizers for NCEP-DOE-R2 and NOAA-CIRES-20CR are just importing the cmorization from NCEP-NCAR-R1. So we felt it would be better to have all three in one PR, since the latter two will not work without the NCEP-NCAR-R1 cmorizer.

valeriupredoi commented:
OK cool, lemme know when you ready so I can start reading :book:

axel-lauer reviewed (changes requested):
I tested the download and formatting scripts for NCEP-DOE-R2, NCEP-NCAR-R1 and NOAA-CIRES-20CR. Those all worked fine after the latest commits.

The CMORized data for NCEP-DOE-R2 and NOAA-CIRES-20CR look good. The following CMORized variables from NCEP-NCAR-R1, however, do not look as I would expect:

* hus: data seem to be in 10-3 (g/kg) instead of the reported unit 1 (kg/kg)
* psl: seems to be in units of hPa instead of the reported Pa
* ta: seems to be in °C instead of the reported K
* ts: seems to be in °C instead of reported K

All additions to the documentation look good to me.

hb326 commented:
I have fixed the problems with the four variables. Still trying to get the Codacy issues sorted, butyou can test the cmorizer again, @axel-lauer.

axel-lauer commented:
Thanks for applying all the fixes. I looked at the CMORized data and NCEP-NCAR-R1 looks fine now. NOAA-CIRES-20CR also looks good but ta from NCEP-DOE-R2 now seems to have been converted from °C to K twice (values now range between 480 and 585 K)...

hb326 commented:
@axel-lauer: I have fixed the problem with the units now. Thanks @schlunma for the great suggestion on how to do it! The units should now be ok for all three datasets. 🤞

axel-lauer reviewed (changes requested):
All 3 CMORizers now produce output that looks reasonable. That's great!

NCEP-NCAR-R1 produces some negative values for prw (water vapor path), which are unphysical. Since the same values are present in the raw files, this should be fine.

I noticed, however, that some variables are missing in the download script for NCEP-NCAR-R1 (I think) and thus skipped when CMORizing the dataset:
* rlut (Amon)
* rlutcs (Amon)
* rsut (Amon)
* rsutcs (Amon)

In addition, monthly mean precipitation (pr) from NCEP-NCAR-R1 is not being CMORized as there seems to be an error in the filename in the cmor_config file (NCEP-NCAR-R1.yml), see suggestion below.

hb326 commented on esmvaltool/cmorizers/data/cmor_config/NCEP-NCAR-R1.yml:
```diff
@@ -0,0 +1,143 @@
+    short_name: pr
+    mip: Amon
+    raw: prate
+    file: 'surface\/prate\.sfc\.mon\.mean\.nc'
```
Well spotted, Axel. I changed the path as suggested.

hb326 commented on esmvaltool/cmorizers/data/cmor_config/NCEP-NCAR-R1.yml:
```diff
@@ -0,0 +1,143 @@
+    short_name: rsutcs
+    mip: Amon
+    raw: csusf
+    file: 'surface\/csusf\.ntat\.mon\.mean\.nc'
```
I added the information about this variable to the download script.

hb326 commented on esmvaltool/cmorizers/data/cmor_config/NCEP-NCAR-R1.yml:
```diff
@@ -0,0 +1,143 @@
+    short_name: rsut
+    mip: Amon
+    raw: uswrf
+    file: 'surface\/uswrf\.ntat\.mon\.mean\.nc'
```
I added the information about this variable to the download script.

hb326 commented on esmvaltool/cmorizers/data/cmor_config/NCEP-NCAR-R1.yml:
```diff
@@ -0,0 +1,143 @@
+    short_name: rlut
+    mip: Amon
+    raw: ulwrf
+    file: 'surface\/ulwrf\.ntat\.mon\.mean\.nc'
```
I added the information about this variable to the download script.

hb326 commented:
@axel-lauer: Very weird that the four radiation variables were not in the downloading script. I have added all four now, and also fixed the file name for the monthly precipitation file.

axel-lauer reviewed (approved):
Thanks @hb326 for applying the suggested changes. Looks all good to me now.

valeriupredoi commented:
cheers Birgit! I can do the technical review then @remi-kazeroni can do the data dance and the final overview from Master Obs Jedi

remi-kazeroni commented:
> cheers Birgit! I can do the technical review then @remi-kazeroni can do the data dance and the final overview from Master Obs Jedi

Thanks for offering @valeriupredoi! If you want, you could focus on the CMORizer scripts (`formatters/datasets/ncep*.py`). After that, I would check the documentation and run the whole thing one last time 🍻

valeriupredoi commented:
> > cheers Birgit! I can do the technical review then @remi-kazeroni can do the data dance and the final overview from Master Obs Jedi
> 
> Thanks for offering @valeriupredoi! If you want, you could focus on the CMORizer scripts (`formatters/datasets/ncep*.py`). After that, I would check the documentation and run the whole thing one last time beers

will do now, bud :beer:

valeriupredoi commented on esmvaltool/cmorizers/data/formatters/datasets/ncep_ncar_r1.py:
```diff
@@ -0,0 +1,169 @@
+
+
+def _fix_units(cube, definition):
+    """Fix issues with the units."""
+    """Exception is 'pr' since the units in the
+       raw file are not recoginzed correctly."""
```
```suggestion
    """
    Fix issues with the units.
    
    Exception is `pr` since the units in the
    raw file are not recognized correctly.
    """
```

valeriupredoi commented on esmvaltool/cmorizers/data/formatters/datasets/ncep_ncar_r1.py:
```diff
@@ -0,0 +1,169 @@
+    """Exception is 'pr' since the units in the
+       raw file are not recoginzed correctly."""
+
+    if cube.var_name not in {'pr'}:
```
```suggestion
    if cube.var_name != 'pr':
```

valeriupredoi reviewed (changes requested):
nice work, guys! Couple very little wiggles from me :beer:

valeriupredoi commented on esmvaltool/cmorizers/data/downloaders/datasets/ncep_doe_r2.py:
```diff
@@ -0,0 +1,47 @@
+
+def download_dataset(config, dataset, dataset_info, start_date, end_date,
+                     overwrite):
+    """Download dataset.
```
```suggestion
    """
    Download dataset.
```

hb326 commented:
Awesome! Thanks @valeriupredoi for the fast check!

valeriupredoi reviewed (approved):
cheers for plugging in the changes so quickly @hb326 :beer: Looking good to me!

remi-kazeroni reviewed (commented):
Thanks a lot for your contribution @hb326! I ran all 3 downloaders and CMORizer scripts successfully, as well as the recipe_check_obs. All looks good to me regarding the code and the data. I have moved the data of these 3 datasets into our RAWOBS and OBS directories. I just have a few minor suggestions to shorten a bit the recipe_check_obs entries and align the instructions for manual downloading with what is done in the automatic scripts. Feel free to take these into account. Apart from that, this PR looks ready to be merged 👍

hb326 commented on esmvaltool/cmorizers/data/formatters/datasets/noaa_cires_20cr.py:
```diff
@@ -0,0 +1,30 @@
+Download and processing instructions
+    To facilitate the download, the links to the ftp server are provided.
+
+    ftp://ftp.cdc.noaa.gov/Projects/20thC_ReanV2/Monthlies/
```
I corrected the path to that it is the same in the formatter and downloader script.

hb326 commented on esmvaltool/cmorizers/data/formatters/datasets/ncep_doe_r2.py:
```diff
@@ -0,0 +1,39 @@
+    20220906
+
+Download and processing instructions
+    To facilitate the download, the links to the ftp server are provided.
```
I corrected the information.

hb326 commented on esmvaltool/cmorizers/data/datasets.yml:
```diff
@@ -872,16 +872,41 @@ datasets:
       save the data in two subdirectories in input_dir_path.
       Subdirectory pressure/:
-        ftp://ftp.cdc.noaa.gov/Datasets/ncep.reanalysis.derived/pressure/
+        ftp://ftp.cdc.noaa.gov/Projects/Datasets/ncep.reanalysis2/Monthlies/
```
Yes, well spotted @remi-kazeroni!
I have corrected the file path now.

remi-kazeroni commented on esmvaltool/cmorizers/data/formatters/datasets/ncep_doe_r2.py:
```diff
@@ -0,0 +1,39 @@
+    20220906
+
+Download and processing instructions
+    To facilitate the download, the links to the ftp server are provided.
+    Since the filenames are sometimes identical across different
+    save the data in two subdirectories in input_dir_path.
+    Subdirectory pressure/:
+      ftp://ftp.cdc.noaa.gov/Projects/Datasets/ncep.reanalysis2/Monthlies/
+      pressure/
+        rhum.mon.mean.nc
+        air.mon.mean.nc
+
+    Subdirectory surface/:
+      ftp://ftp.cdc.noaa.gov/Datasets/ncep.reanalysis2/Monthlies/gaussian_grid
+        tcdc.eatm.mon.mean.nc
+      ftp://ftp.cdc.noaa.gov/Datasets/ncep.reanalysis2/Monthlies/surface
+        pr_wtr.eatm.mon.mean.nc
+
+    #Select the section "Pressure" and "Surface" and download the variables
+    #listed below. Since raw data on pressure levels and for surface have the
+    #same file and variable name, save the data in two different subdirectories
+    #"press" and "surf" in input_dir_path.
```
I think we forgot about this file (some remaining copy-pasting, it seems). Shouldn't this be like?
```suggestion
    To facilitate the download, the links to the https server are provided.
    
      https://downloads.psl.noaa.gov/Datasets/ncep.reanalysis2/Monthlies/
      pressure/
        rhum.mon.mean.nc
        air.mon.mean.nc
      https://downloads.psl.noaa.gov/Datasets/ncep.reanalysis2/Monthlies/
      gaussian_grid/
        tcdc.eatm.mon.mean.nc
      https://downlooads.psl.noaa.gov/Datasets/ncep.reanalysis2/Monthlies/
      surface/
        pr_wtr.eatm.mon.mean.nc

```

remi-kazeroni reviewed (commented):
Thanks for making the changes @hb326! I think we forgot to check the documentation in one file. Feel free to take my suggestion into account and I think we are good for the merge 👍

hb326 commented:
> Thanks for making the changes @hb326! I think we forgot to check the documentation in one file. Feel free to take my suggestion into account and I think we are good for the merge 👍

@remi-kazeroni: yes, you are right! It should have read like you suggested. I changed the file just now.

remi-kazeroni reviewed (approved):
Thanks a lot for your work @hb326 and everyone for the reviews! Let's take care of PR #2932 to update recipes accordingly before merging this PR. The raw and cmorized data are available on the common RAWOBS and OBS directories on Levante.
